common/flatpak-run-wayland.c

```diff
@@ -0,0 +1,218 @@
+
+  display = wl_display_connect (NULL);
+  if (!display)
+    return FALSE;
+
+  registry = wl_display_get_registry (display);
+  wl_registry_add_listener (registry, &registry_listener,
+                            &security_context_manager);
+  ret = wl_display_roundtrip (display);
+  wl_registry_destroy (registry);
+  if (ret < 0)
+    goto out;
+
+  if (!security_context_manager) {
+    *available_out = FALSE;
+    goto out;
+  }
+
+  socket_path = create_wl_socket ("wayland-XXXXXX");
+  if (!socket_path)
+    goto out;
+
+  unlink (socket_path);
+
+  listen_fd = socket (AF_UNIX, SOCK_STREAM, 0);
+  if (listen_fd < 0)
+    goto out;
+
+  sockaddr.sun_family = AF_UNIX;
+  snprintf (sockaddr.sun_path, sizeof(sockaddr.sun_path), "%s", socket_path);
+  if (bind (listen_fd, (struct sockaddr *) &sockaddr, sizeof (sockaddr)) != 0)
+    goto out;
+
+  if (listen (listen_fd, 0) != 0)
+    goto out;
+
+  sync_fd = flatpak_bwrap_add_sync_fd (bwrap);
+  if (sync_fd < 0)
+    goto out;
+
+  security_context = wp_security_context_manager_v1_create_listener (security_context_manager,
+                                                                     listen_fd,
+                                                                     sync_fd);
+  wp_security_context_v1_set_sandbox_engine (security_context, "flatpak");
+  wp_security_context_v1_set_app_id (security_context, app_id);
+  wp_security_context_v1_set_instance_id (security_context, instance_id);
+  wp_security_context_v1_commit (security_context);
+  wp_security_context_v1_destroy (security_context);
+  if (wl_display_roundtrip (display) < 0)
+    goto out;
+
+  flatpak_bwrap_add_args (bwrap,
+                          "--ro-bind", socket_path, "/run/flatpak/wayland-0",
+                          NULL);
+  flatpak_bwrap_set_env (bwrap, "WAYLAND_DISPLAY", "/run/flatpak/wayland-0", TRUE);
+
+  res = TRUE;
+
+out:
+  if (listen_fd >= 0)
+    close (listen_fd);
+  if (security_context_manager)
+    wp_security_context_manager_v1_destroy (security_context_manager);
+  wl_display_disconnect (display);
+  return res;
+}
+
+#endif /* ENABLE_WAYLAND_SECURITY_CONTEXT */
+
+/**
+ * flatpak_run_add_wayland_args:
+ *
+ * Returns: %TRUE if a Wayland socket was found.
+ */
+gboolean
+flatpak_run_add_wayland_args (FlatpakBwrap *bwrap,
+                              const char   *app_id,
+                              const char   *instance_id)
+{
+  const char *wayland_display;
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  g_autofree char *wayland_socket = NULL;
+  g_autofree char *sandbox_wayland_socket = NULL;
+  gboolean res = FALSE;
+  struct stat statbuf;
+
+#ifdef ENABLE_WAYLAND_SECURITY_CONTEXT
+  gboolean security_context_available = FALSE;
+  if (flatpak_run_add_wayland_security_context_args (bwrap, app_id, instance_id,
+                                                     &security_context_available))
+    return TRUE;
+  if (security_context_available)
+    return FALSE;
```

Comment:
Why security context create failed will return FALSE?
Though `security_context_available` is true, if still using wayland mode, what will happen?.

Reply:
The goal here is to bail out if the security context protocol is supported by the compositor but we didn't manage to set it up. If the security context protocol is not supported, we want to carry on.

Reply:
No, continuing would be insecure, potentially running the client without a security context.

I don't think Flatpak can run inside Flatpak, so this doesn't really matter.